pkg/resource/deploy/step_generator.go

```diff
@@ -555,7 +555,96 @@ func (sg *stepGenerator) GenerateDeletes() []Step {
 			}
 		}
 	}
-	return dels
+
+	if len(targets) > 0 {
+		logging.V(7).Infof("Planner was asked to only delete '%v'", targets)
+
+		resourcesToDelete, res := sg.computeResourcesToDelete(targets)
+		if res != nil {
+			return nil, res
+		}
+
+		keys := []resource.URN{}
+		for k := range resourcesToDelete {
+			keys = append(keys, k)
+		}
+
+		logging.V(7).Infof("Planner will delete all of '%v'", keys)
+
+		filtered := []Step{}
+		for _, step := range dels {
+			if _, has := resourcesToDelete[step.URN()]; has {
+				filtered = append(filtered, step)
+			}
+		}
+
+		dels = filtered
+	}
+
+	return dels, nil
+}
+
+func (sg *stepGenerator) computeResourcesToDelete(roots []resource.URN) (map[resource.URN]bool, result.Result) {
```

Comment:
Why isn't this using the pre-existing code that handles delete-before-replace?

Reply:
so, examining the existing [calculateDependentResources](https://github.com/pulumi/pulumi/blob/1387afec8fbed2b74f7cf3b0d06bf24f2ce22f66/pkg/resource/deploy/step_generator.go#L889) logic, it appeared to not actually provide value as far as i could see.  It was already doing something similar to what i do here (looking at the dep-graph to determine downstream affected resources).  But i couldn't identify a scenario where it was valuable to only work with these resources.

For example, in the tests here: https://github.com/pulumi/pulumi/blob/master/pkg/engine/lifecycle_test.go#L2625

```
	//             A
	//    _________|_________
	//    B        C        D
	//          ___|___  ___|___
	//          E  F  G  H  I  J
	//             |__|
	//             K  L
	//
	// For a given resource R in (A, C, D):
	// - R will be the provider for its first dependent
	// - A change to R will require that its second dependent be replaced
	// - A change to R will not require that its third dependent be replaced
	//
	// In addition, K will have a requires-replacement property that depends on both F and G, and
	// L will have a normal property that depends on both F and G.
```

the code that handles delete-before-replace identifies `B, C, E, F, and K` as needing replacement for when 'A' is affected.  However, if 'A' is deleted, we still need to deal with 'D' still having a dep on 'A'.  If we do not handle it, then we'll have a dangling ref.

It simply seemed sensible to me (but it sounds like i missed something) that asking to destroy a resource A would necessarily destroy what was necessary in order to be able to properly get rid of that resource itself.  

Could you clarify how you intended the existing DBR logic be used, and how it should handle a case like this?

Reply:
> However, if 'A' is deleted, we still need to deal with 'D' still having a dep on 'A'. If we do not handle it, then we'll have a dangling ref.

We should deal with this by rewriting A's dependencies ala refresh.

> It simply seemed sensible to me (but it sounds like i missed something) that asking to destroy a resource A would necessarily destroy what was necessary in order to be able to properly get rid of that resource itself.

This is exactly the purpose of `calculateDependentReplacements`. That function goes out of its way to ensure that the smallest set of dependents are deleted. There is a largish comment inside that function that explains its behavior and the rationale behind it: https://github.com/pulumi/pulumi/blob/1387afec8fbed2b74f7cf3b0d06bf24f2ce22f66/pkg/resource/deploy/step_generator.go#L890-L912

The code you have here will not achieve the same effect. It will delete potentially many more resources than would be deleted by a typical delete-before-replace. DBR used to use almost exactly the algorithm you've implemented, and that proved to be an unsatisfactory user experience.

In addition, I think that we should just ban the deletion of resources that have children for the time being. We can fix that when we resolve https://github.com/pulumi/pulumi/issues/2327.

Reply:
@pgavlin sounds good.  let me try that out.

Reply:
> In addition, I think that we should just ban the deletion of resources that have children for the time being. We can fix that when we resolve #2327.

Hrmm.. That seems like a heavy hammer.  It doesn't seem strange to me to propagate from parents to children.  After all, that's what we would force users to do to get through this situation.  Except now we can do that for them rather than making them take all the manual steps.

Is there a problem with supporting this scenario up front?  Thanks!

Reply:
> Is there a problem with supporting this scenario up front? Thanks!

Yes: I'd prefer that the code that performs the deletion of a specific resource retain the same behavior that we have with DBR. I don't think that it's going to be a good UX to require that children of a parent are deleted if the parent is deleted (the dependency there isn't "real"; we only have that requirement so the state file remains well-formed).

Reply:
alrighty.  will remove that code :) 

Note, i'm not sure i agree with this:

> I don't think that it's going to be a good UX to require that children of a parent are deleted if the parent is deleted (the dependency there isn't "real"; we only have that requirement so the state file remains well-formed).

Regardless of what we want or not, that restriction *does* exist on the user today.  They will not be able to delete a parent before deleting all children.  So removing this facility doesn't change anything, it just makes the scenario harder.  That said, i imagine it's the 99% scenario to want to delete the leaf-resource anyways.  so removing this support seems fine :)